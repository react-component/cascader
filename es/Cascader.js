import _extends from 'babel-runtime/helpers/extends';
import _objectWithoutProperties from 'babel-runtime/helpers/objectWithoutProperties';
import _toConsumableArray from 'babel-runtime/helpers/toConsumableArray';
import _classCallCheck from 'babel-runtime/helpers/classCallCheck';
import _createClass from 'babel-runtime/helpers/createClass';
import _possibleConstructorReturn from 'babel-runtime/helpers/possibleConstructorReturn';
import _inherits from 'babel-runtime/helpers/inherits';
import React, { Component, cloneElement } from 'react';
import PropTypes from 'prop-types';
import Trigger from 'rc-trigger';
import warning from 'warning';
import KeyCode from 'rc-util/es/KeyCode';
import arrayTreeFilter from 'array-tree-filter';
import shallowEqualArrays from 'shallow-equal/arrays';
import { polyfill } from 'react-lifecycles-compat';
import Menus from './Menus';

var BUILT_IN_PLACEMENTS = {
  bottomLeft: {
    points: ['tl', 'bl'],
    offset: [0, 4],
    overflow: {
      adjustX: 1,
      adjustY: 1,
    },
  },
  topLeft: {
    points: ['bl', 'tl'],
    offset: [0, -4],
    overflow: {
      adjustX: 1,
      adjustY: 1,
    },
  },
  bottomRight: {
    points: ['tr', 'br'],
    offset: [0, 4],
    overflow: {
      adjustX: 1,
      adjustY: 1,
    },
  },
  topRight: {
    points: ['br', 'tr'],
    offset: [0, -4],
    overflow: {
      adjustX: 1,
      adjustY: 1,
    },
  },
};

var Cascader = (function(_Component) {
  _inherits(Cascader, _Component);

  function Cascader(props) {
    _classCallCheck(this, Cascader);

    var _this = _possibleConstructorReturn(
      this,
      (Cascader.__proto__ || Object.getPrototypeOf(Cascader)).call(this, props),
    );

    _this.setPopupVisible = function(popupVisible) {
      if (!('popupVisible' in _this.props)) {
        _this.setState({ popupVisible: popupVisible });
      }
      // sync activeValue with value when panel open
      if (popupVisible && !_this.state.popupVisible) {
        _this.setState({
          activeValue: _this.state.value,
        });
      }
      _this.props.onPopupVisibleChange(popupVisible);
    };

    _this.handleChange = function(options, setProps, e) {
      if (e.type !== 'keydown' || e.keyCode === KeyCode.ENTER) {
        _this.props.onChange(
          options.map(function(o) {
            return o[_this.getFieldName('value')];
          }),
          options,
        );
        _this.setPopupVisible(setProps.visible);
      }
    };

    _this.handlePopupVisibleChange = function(popupVisible) {
      _this.setPopupVisible(popupVisible);
    };

    _this.handleMenuSelect = function(targetOption, menuIndex, e) {
      // Keep focused state for keyboard support
      var triggerNode = _this.trigger.getRootDomNode();
      if (triggerNode && triggerNode.focus) {
        triggerNode.focus();
      }
      var _this$props = _this.props,
        changeOnSelect = _this$props.changeOnSelect,
        loadData = _this$props.loadData,
        expandTrigger = _this$props.expandTrigger;

      if (!targetOption || targetOption.disabled) {
        return;
      }
      var activeValue = _this.state.activeValue;

      activeValue = activeValue.slice(0, menuIndex + 1);
      activeValue[menuIndex] = targetOption[_this.getFieldName('value')];
      var activeOptions = _this.getActiveOptions(activeValue);
      if (
        targetOption.isLeaf === false &&
        !targetOption[_this.getFieldName('children')] &&
        loadData
      ) {
        if (changeOnSelect) {
          _this.handleChange(activeOptions, { visible: true }, e);
        }
        _this.setState({ activeValue: activeValue });
        loadData(activeOptions);
        return;
      }
      var newState = {};
      if (
        !targetOption[_this.getFieldName('children')] ||
        !targetOption[_this.getFieldName('children')].length
      ) {
        _this.handleChange(activeOptions, { visible: false }, e);
        // set value to activeValue when select leaf option
        newState.value = activeValue;
        // add e.type judgement to prevent `onChange` being triggered by mouseEnter
      } else if (changeOnSelect && (e.type === 'click' || e.type === 'keydown')) {
        if (expandTrigger === 'hover') {
          _this.handleChange(activeOptions, { visible: false }, e);
        } else {
          _this.handleChange(activeOptions, { visible: true }, e);
        }
        // set value to activeValue on every select
        newState.value = activeValue;
      }
      newState.activeValue = activeValue;
      //  not change the value by keyboard
      if ('value' in _this.props || (e.type === 'keydown' && e.keyCode !== KeyCode.ENTER)) {
        delete newState.value;
      }
      _this.setState(newState);
    };

    _this.handleItemDoubleClick = function() {
      var changeOnSelect = _this.props.changeOnSelect;

      if (changeOnSelect) {
        _this.setPopupVisible(false);
      }
    };

    _this.handleKeyDown = function(e) {
      var children = _this.props.children;
      // https://github.com/ant-design/ant-design/issues/6717
      // Don't bind keyboard support when children specify the onKeyDown

      if (children && children.props.onKeyDown) {
        children.props.onKeyDown(e);
        return;
      }
      var activeValue = [].concat(_toConsumableArray(_this.state.activeValue));
      var currentLevel = activeValue.length - 1 < 0 ? 0 : activeValue.length - 1;
      var currentOptions = _this.getCurrentLevelOptions();
      var currentIndex = currentOptions
        .map(function(o) {
          return o[_this.getFieldName('value')];
        })
        .indexOf(activeValue[currentLevel]);
      if (
        e.keyCode !== KeyCode.DOWN &&
        e.keyCode !== KeyCode.UP &&
        e.keyCode !== KeyCode.LEFT &&
        e.keyCode !== KeyCode.RIGHT &&
        e.keyCode !== KeyCode.ENTER &&
        e.keyCode !== KeyCode.BACKSPACE &&
        e.keyCode !== KeyCode.ESC
      ) {
        return;
      }
      // Press any keys above to reopen menu
      if (
        !_this.state.popupVisible &&
        e.keyCode !== KeyCode.BACKSPACE &&
        e.keyCode !== KeyCode.LEFT &&
        e.keyCode !== KeyCode.RIGHT &&
        e.keyCode !== KeyCode.ESC
      ) {
        _this.setPopupVisible(true);
        return;
      }
      if (e.keyCode === KeyCode.DOWN || e.keyCode === KeyCode.UP) {
        e.preventDefault();
        var nextIndex = currentIndex;
        if (nextIndex !== -1) {
          if (e.keyCode === KeyCode.DOWN) {
            nextIndex += 1;
            nextIndex = nextIndex >= currentOptions.length ? 0 : nextIndex;
          } else {
            nextIndex -= 1;
            nextIndex = nextIndex < 0 ? currentOptions.length - 1 : nextIndex;
          }
        } else {
          nextIndex = 0;
        }
        activeValue[currentLevel] = currentOptions[nextIndex][_this.getFieldName('value')];
      } else if (e.keyCode === KeyCode.LEFT || e.keyCode === KeyCode.BACKSPACE) {
        e.preventDefault();
        activeValue.splice(activeValue.length - 1, 1);
      } else if (e.keyCode === KeyCode.RIGHT) {
        e.preventDefault();
        if (
          currentOptions[currentIndex] &&
          currentOptions[currentIndex][_this.getFieldName('children')]
        ) {
          activeValue.push(
            currentOptions[currentIndex][_this.getFieldName('children')][0][
              _this.getFieldName('value')
            ],
          );
        }
      } else if (e.keyCode === KeyCode.ESC) {
        _this.setPopupVisible(false);
        return;
      }
      if (!activeValue || activeValue.length === 0) {
        _this.setPopupVisible(false);
      }
      var activeOptions = _this.getActiveOptions(activeValue);
      var targetOption = activeOptions[activeOptions.length - 1];
      _this.handleMenuSelect(targetOption, activeOptions.length - 1, e);

      if (_this.props.onKeyDown) {
        _this.props.onKeyDown(e);
      }
    };

    _this.saveTrigger = function(node) {
      _this.trigger = node;
    };

    var initialValue = [];
    if ('value' in props) {
      initialValue = props.value || [];
    } else if ('defaultValue' in props) {
      initialValue = props.defaultValue || [];
    }

    warning(
      !('filedNames' in props),
      '`filedNames` of Cascader is a typo usage and deprecated, please use `fieldNames` instead.',
    );

    _this.state = {
      popupVisible: props.popupVisible,
      activeValue: initialValue,
      value: initialValue,
      prevProps: props,
    };
    _this.defaultFieldNames = { label: 'label', value: 'value', children: 'children' };
    return _this;
  }

  _createClass(
    Cascader,
    [
      {
        key: 'getPopupDOMNode',
        value: function getPopupDOMNode() {
          return this.trigger.getPopupDomNode();
        },
      },
      {
        key: 'getFieldName',
        value: function getFieldName(name) {
          var defaultFieldNames = this.defaultFieldNames;
          var _props = this.props,
            fieldNames = _props.fieldNames,
            filedNames = _props.filedNames;

          if ('filedNames' in this.props) {
            return filedNames[name] || defaultFieldNames[name]; // For old compatibility
          }
          return fieldNames[name] || defaultFieldNames[name];
        },
      },
      {
        key: 'getFieldNames',
        value: function getFieldNames() {
          var _props2 = this.props,
            fieldNames = _props2.fieldNames,
            filedNames = _props2.filedNames;

          if ('filedNames' in this.props) {
            return filedNames; // For old compatibility
          }
          return fieldNames;
        },
      },
      {
        key: 'getCurrentLevelOptions',
        value: function getCurrentLevelOptions() {
          var _this2 = this;

          var _props$options = this.props.options,
            options = _props$options === undefined ? [] : _props$options;
          var _state$activeValue = this.state.activeValue,
            activeValue = _state$activeValue === undefined ? [] : _state$activeValue;

          var result = arrayTreeFilter(
            options,
            function(o, level) {
              return o[_this2.getFieldName('value')] === activeValue[level];
            },
            { childrenKeyName: this.getFieldName('children') },
          );
          if (result[result.length - 2]) {
            return result[result.length - 2][this.getFieldName('children')];
          }
          return [].concat(_toConsumableArray(options)).filter(function(o) {
            return !o.disabled;
          });
        },
      },
      {
        key: 'getActiveOptions',
        value: function getActiveOptions(activeValue) {
          var _this3 = this;

          return arrayTreeFilter(
            this.props.options || [],
            function(o, level) {
              return o[_this3.getFieldName('value')] === activeValue[level];
            },
            { childrenKeyName: this.getFieldName('children') },
          );
        },
      },
      {
        key: 'render',
        value: function render() {
          var _props3 = this.props,
            prefixCls = _props3.prefixCls,
            transitionName = _props3.transitionName,
            popupClassName = _props3.popupClassName,
            _props3$options = _props3.options,
            options = _props3$options === undefined ? [] : _props3$options,
            disabled = _props3.disabled,
            builtinPlacements = _props3.builtinPlacements,
            popupPlacement = _props3.popupPlacement,
            children = _props3.children,
            restProps = _objectWithoutProperties(_props3, [
              'prefixCls',
              'transitionName',
              'popupClassName',
              'options',
              'disabled',
              'builtinPlacements',
              'popupPlacement',
              'children',
            ]);
          // Did not show popup when there is no options

          var menus = React.createElement('div', null);
          var emptyMenuClassName = '';
          if (options && options.length > 0) {
            menus = React.createElement(
              Menus,
              _extends({}, this.props, {
                fieldNames: this.getFieldNames(),
                defaultFieldNames: this.defaultFieldNames,
                activeValue: this.state.activeValue,
                onSelect: this.handleMenuSelect,
                onItemDoubleClick: this.handleItemDoubleClick,
                visible: this.state.popupVisible,
              }),
            );
          } else {
            emptyMenuClassName = ' ' + prefixCls + '-menus-empty';
          }
          return React.createElement(
            Trigger,
            _extends(
              {
                ref: this.saveTrigger,
              },
              restProps,
              {
                options: options,
                disabled: disabled,
                popupPlacement: popupPlacement,
                builtinPlacements: builtinPlacements,
                popupTransitionName: transitionName,
                action: disabled ? [] : ['click'],
                popupVisible: disabled ? false : this.state.popupVisible,
                onPopupVisibleChange: this.handlePopupVisibleChange,
                prefixCls: prefixCls + '-menus',
                popupClassName: popupClassName + emptyMenuClassName,
                popup: menus,
              },
            ),
            cloneElement(children, {
              onKeyDown: this.handleKeyDown,
              tabIndex: disabled ? undefined : 0,
            }),
          );
        },
      },
    ],
    [
      {
        key: 'getDerivedStateFromProps',
        value: function getDerivedStateFromProps(nextProps, prevState) {
          var _prevState$prevProps = prevState.prevProps,
            prevProps = _prevState$prevProps === undefined ? {} : _prevState$prevProps;

          var newState = {
            prevProps: nextProps,
          };

          if ('value' in nextProps && !shallowEqualArrays(prevProps.value, nextProps.value)) {
            newState.value = nextProps.value || [];

            // allow activeValue diff from value
            // https://github.com/ant-design/ant-design/issues/2767
            if (!('loadData' in nextProps)) {
              newState.activeValue = nextProps.value || [];
            }
          }
          if ('popupVisible' in nextProps) {
            newState.popupVisible = nextProps.popupVisible;
          }

          return newState;
        },
      },
    ],
  );

  return Cascader;
})(Component);

Cascader.defaultProps = {
  onChange: function onChange() {},
  onPopupVisibleChange: function onPopupVisibleChange() {},

  disabled: false,
  transitionName: '',
  prefixCls: 'rc-cascader',
  popupClassName: '',
  popupPlacement: 'bottomLeft',
  builtinPlacements: BUILT_IN_PLACEMENTS,
  expandTrigger: 'click',
  fieldNames: { label: 'label', value: 'value', children: 'children' },
  expandIcon: '>',
};

Cascader.propTypes = {
  value: PropTypes.array,
  defaultValue: PropTypes.array,
  options: PropTypes.array.isRequired,
  onChange: PropTypes.func,
  onPopupVisibleChange: PropTypes.func,
  popupVisible: PropTypes.bool,
  disabled: PropTypes.bool,
  transitionName: PropTypes.string,
  popupClassName: PropTypes.string,
  popupPlacement: PropTypes.string,
  prefixCls: PropTypes.string,
  dropdownMenuColumnStyle: PropTypes.object,
  builtinPlacements: PropTypes.object,
  loadData: PropTypes.func,
  changeOnSelect: PropTypes.bool,
  children: PropTypes.node,
  onKeyDown: PropTypes.func,
  expandTrigger: PropTypes.string,
  fieldNames: PropTypes.object,
  filedNames: PropTypes.object, // typo but for compatibility
  expandIcon: PropTypes.node,
  loadingIcon: PropTypes.node,
};

polyfill(Cascader);

export default Cascader;
