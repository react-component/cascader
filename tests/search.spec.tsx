import { fireEvent, render } from '@testing-library/react';
import KeyCode from 'rc-util/lib/KeyCode';
import { resetWarned } from 'rc-util/lib/warning';
import React from 'react';
import Cascader from '../src';
import { optionsForActiveMenuItems } from './demoOptions';
import type { ReactWrapper } from './enzyme';
import { mount } from './enzyme';

describe('Cascader.Search', () => {
  function doSearch(wrapper: ReactWrapper, search: string) {
    wrapper.find('input').simulate('change', {
      target: {
        value: search,
      },
    });
  }

  const options = [
    {
      label: 'Label Light',
      value: 'light',
    },
    {
      label: 'Label Bamboo',
      value: 'bamboo',
      children: [
        {
          label: 'Label Little',
          value: 'little',
          children: [
            {
              label: 'Toy Fish',
              value: 'fish',
              // Leave a empty children here. But cascader should think this is a leaf node.
              children: [],
            },
            {
              label: 'Toy Cards',
              value: 'cards',
            },
          ],
        },
      ],
    },
  ];

  it('default search', () => {
    const onSearch = jest.fn();
    const onChange = jest.fn();
    const wrapper = mount(
      <Cascader options={options} onChange={onChange} onSearch={onSearch} open showSearch />,
    );

    // Leaf
    doSearch(wrapper, 'toy');
    let itemList = wrapper.find('div.rc-cascader-menu-item-content');
    expect(itemList).toHaveLength(2);
    expect(itemList.at(0).text()).toEqual('Label Bamboo / Label Little / Toy Fish');
    expect(itemList.at(1).text()).toEqual('Label Bamboo / Label Little / Toy Cards');
    expect(onSearch).toHaveBeenCalledWith('toy');

    // Parent
    doSearch(wrapper, 'Label Little');
    itemList = wrapper.find('div.rc-cascader-menu-item-content');
    expect(itemList).toHaveLength(2);
    expect(itemList.at(0).text()).toEqual('Label Bamboo / Label Little / Toy Fish');
    expect(itemList.at(1).text()).toEqual('Label Bamboo / Label Little / Toy Cards');
    expect(onSearch).toHaveBeenCalledWith('Label Little');

    // Change
    wrapper.clickOption(0, 0);
    expect(onChange).toHaveBeenCalledWith(['bamboo', 'little', 'fish'], expect.anything());
  });

  it('changeOnSelect', () => {
    const onChange = jest.fn();
    const wrapper = mount(
      <Cascader options={options} onChange={onChange} open showSearch changeOnSelect />,
    );

    // Leaf
    doSearch(wrapper, 'Label Little');
    const itemList = wrapper.find('div.rc-cascader-menu-item-content');
    expect(itemList).toHaveLength(3);
    expect(itemList.at(0).text()).toEqual('Label Bamboo / Label Little');
    expect(itemList.at(1).text()).toEqual('Label Bamboo / Label Little / Toy Fish');
    expect(itemList.at(2).text()).toEqual('Label Bamboo / Label Little / Toy Cards');

    // Should not expandable
    expect(wrapper.exists('.rc-cascader-menu-item-expand-icon')).toBeFalsy();

    // Trigger onChange
    wrapper.find('input').simulate('keyDown', { which: KeyCode.DOWN });
    wrapper.find('input').simulate('keyDown', { which: KeyCode.ENTER });
    expect(onChange).toHaveBeenCalledWith(['bamboo', 'little'], expect.anything());
  });

  it('sort', () => {
    const wrapper = mount(
      <Cascader
        options={options}
        open
        showSearch={{
          sort: (pathA, pathB) => {
            const finalA = pathA[pathA.length - 1];
            const finalB = pathB[pathB.length - 1];

            // this value is string
            if ((finalA.value as any) < (finalB.value as any)) {
              return -1;
            }
            return 1;
          },
        }}
      />,
    );

    doSearch(wrapper, 'toy');
    const itemList = wrapper.find('div.rc-cascader-menu-item-content');
    expect(itemList).toHaveLength(2);
    expect(itemList.at(0).text()).toEqual('Label Bamboo / Label Little / Toy Cards');
    expect(itemList.at(1).text()).toEqual('Label Bamboo / Label Little / Toy Fish');
  });

  it('limit', () => {
    const wrapper = mount(
      <Cascader
        options={options}
        open
        showSearch={{
          limit: 1,
        }}
      />,
    );

    doSearch(wrapper, 'toy');
    const itemList = wrapper.find('div.rc-cascader-menu-item-content');
    expect(itemList).toHaveLength(1);
    expect(itemList.at(0).text()).toEqual('Label Bamboo / Label Little / Toy Fish');
  });

  it('render', () => {
    const wrapper = mount(
      <Cascader
        options={options}
        open
        showSearch={{
          render: (inputValue, optList, prefixCls) =>
            `${prefixCls}-${inputValue}-${optList.map(opt => opt.value).join('~')}`,
        }}
      />,
    );

    doSearch(wrapper, 'toy');
    const itemList = wrapper.find('div.rc-cascader-menu-item-content');
    expect(itemList).toHaveLength(2);
    expect(itemList.at(0).text()).toEqual('rc-cascader-toy-bamboo~little~fish');
    expect(itemList.at(1).text()).toEqual('rc-cascader-toy-bamboo~little~cards');
  });

  it('not crash when empty', () => {
    const onChange = jest.fn();
    const wrapper = mount(<Cascader options={options} onChange={onChange} showSearch />);
    doSearch(wrapper, 'toy');

    // Selection empty
    wrapper.find('input').simulate('keyDown', { which: KeyCode.ENTER });
    expect(onChange).not.toHaveBeenCalled();

    wrapper.find('input').simulate('keyDown', { which: KeyCode.DOWN });
    wrapper.find('input').simulate('keyDown', { which: KeyCode.ENTER });
    expect(onChange).toHaveBeenCalled();

    // Content empty
    doSearch(wrapper, 'not exist');
    expect(wrapper.exists('.rc-cascader-menu-empty')).toBeTruthy();
  });

  it('warning of negative limit', () => {
    resetWarned();
    const errorSpy = jest.spyOn(console, 'error').mockImplementation(() => {});

    const wrapper = mount(<Cascader options={options} showSearch={{ limit: 0 }} />);

    expect(errorSpy).toHaveBeenCalledWith(
      "Warning: 'limit' of showSearch should be positive number or false.",
    );

    doSearch(wrapper, 'toy');
    expect(wrapper.find('div.rc-cascader-menu-item-content')).toHaveLength(2);

    errorSpy.mockRestore();
  });

  it('onChange should be triggered when click option with changeOnSelect + multiple', () => {
    const onChange = jest.fn();
    const wrapper = mount(
      <Cascader checkable options={options} changeOnSelect onChange={onChange} showSearch />,
    );
    doSearch(wrapper, 'toy');
    wrapper.find('.rc-cascader-menu-item').first().simulate('click');
    wrapper.find('.rc-cascader-menu-item').first().simulate('mousedown');
    expect(onChange).toHaveBeenCalledWith([['bamboo', 'little', 'fish']], expect.anything());

    doSearch(wrapper, 'light');
    wrapper.find('.rc-cascader-menu-item').first().simulate('click');
    wrapper.find('.rc-cascader-menu-item').first().simulate('mousedown');
    expect(onChange).toHaveBeenCalledWith(
      [['bamboo', 'little', 'fish'], ['light']],
      expect.anything(),
    );
  });

  it('onChange should be triggered when click option with multiple', () => {
    const onChange = jest.fn();
    const wrapper = mount(<Cascader checkable options={options} onChange={onChange} showSearch />);
    doSearch(wrapper, 'toy');
    wrapper.find('.rc-cascader-menu-item').first().simulate('click');
    wrapper.find('.rc-cascader-menu-item').first().simulate('mousedown');
    expect(onChange).toHaveBeenCalledWith([['bamboo', 'little', 'fish']], expect.anything());

    doSearch(wrapper, 'light');
    wrapper.find('.rc-cascader-menu-item').first().simulate('click');
    wrapper.find('.rc-cascader-menu-item').first().simulate('mousedown');
    expect(onChange).toHaveBeenCalledWith(
      [['bamboo', 'little', 'fish'], ['light']],
      expect.anything(),
    );
  });

  it('should not crash when exist options with same value on different levels', () => {
    const wrapper = mount(<Cascader options={optionsForActiveMenuItems} />);

    doSearch(wrapper, '1');
    wrapper.find('.rc-cascader-menu-item').first().simulate('click');
    doSearch(wrapper, '1');
  });

  it('should correct render Cascader with same field name of label and value', () => {
    const customOptions = [
      {
        name: 'Zhejiang',
        children: [
          {
            name: 'Hangzhou',
            children: [
              {
                name: 'West Lake',
              },
              {
                name: 'Xia Sha',
                disabled: true,
              },
            ],
          },
        ],
      },
    ];
    const wrapper = mount(
      <Cascader
        options={customOptions}
        fieldNames={{ label: 'name', value: 'name' }}
        showSearch={{
          filter: (inputValue, path) =>
            path.some(option => option.name.toLowerCase().indexOf(inputValue.toLowerCase()) > -1),
        }}
      />,
    );
    wrapper.find('input').simulate('change', { target: { value: 'z' } });
    expect(wrapper.render()).toMatchSnapshot();
  });

  // https://github.com/ant-design/ant-design/issues/41810
  it('not back to options when selected', () => {
    const { container } = render(<Cascader options={options} showSearch />);

    // Search
    fireEvent.change(container.querySelector('input') as HTMLElement, {
      target: {
        value: 'bamboo',
      },
    });

    // Click
    fireEvent.click(document.querySelector('.rc-cascader-menu-item-content') as HTMLElement);
    expect(document.querySelector('.rc-cascader-dropdown-hidden')).toBeTruthy();
    expect(document.querySelector('.rc-cascader-menu-item-content')?.textContent).toBe(
      'Label Bamboo / Label Little / Toy Fish',
    );
  });

  it('autoClearSearchValue={false} should be worked', () => {
    const wrapper = mount(
      <Cascader options={options} showSearch checkable autoClearSearchValue={false} />,
    );

    // Search
    wrapper.find('input').simulate('change', { target: { value: 'bamboo' } });

    // Click
    wrapper.find('.rc-cascader-checkbox').first().simulate('click');
    expect(wrapper.find('input').prop('value')).toEqual('bamboo');
  });

  it('disabled path should not search', () => {
    const { container } = render(
      <Cascader
        open
        searchValue="little"
        options={[
          {
            label: 'bamboo',
            value: 'bamboo',
            disabled: true,
            children: [
              {
                label: 'little',
                value: 'little',
              },
            ],
          },
        ]}
      />,
    );

    expect(container.querySelectorAll('.rc-cascader-menu-item')).toHaveLength(1);
    expect(container.querySelectorAll('.rc-cascader-menu-item-disabled')).toHaveLength(1);
    expect(container.querySelector('.rc-cascader-menu-item-disabled')?.textContent).toEqual(
      'bamboo / little',
    );
  });
  it('Should optionRender work', () => {
    const { container, rerender } = render(
      <Cascader
        open
        options={[{ label: 'bamboo', value: 'bamboo' }]}
        optionRender={option => `${option.label} - test`}
      />,
    );
    expect(container.querySelector('.rc-cascader-menu-item-content')?.innerHTML).toEqual(
      'bamboo - test',
    );
    rerender(
      <Cascader
        open
        options={[{ label: 'bamboo', disabled: true, value: 'bamboo' }]}
        optionRender={option => JSON.stringify(option)}
      />,
    );
    expect(container.querySelector('.rc-cascader-menu-item-content')?.innerHTML).toEqual(
      '{"label":"bamboo","disabled":true,"value":"bamboo"}',
    );
  });
});
