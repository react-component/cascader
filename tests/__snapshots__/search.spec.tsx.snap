// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cascader.Search should correct render Cascader with same field name of label and value 1`] = `
<div
  class="rc-cascader rc-cascader-single rc-cascader-open rc-cascader-show-search"
>
  <div
    class="rc-cascader-selector"
  >
    <span
      class="rc-cascader-selection-search"
    >
      <input
        aria-autocomplete="list"
        aria-controls="rc_select_TEST_OR_SSR_list"
        aria-expanded="true"
        aria-haspopup="listbox"
        aria-owns="rc_select_TEST_OR_SSR_list"
        autocomplete="off"
        class="rc-cascader-selection-search-input"
        id="rc_select_TEST_OR_SSR"
        role="combobox"
        type="search"
        value="z"
      />
    </span>
    <span
      class="rc-cascader-selection-placeholder"
      style="visibility: hidden;"
    />
  </div>
  <div
    class="rc-cascader-dropdown rc-cascader-dropdown-placement-bottomLeft"
    style="--arrow-x: 0px; --arrow-y: 0px; left: -1000vw; top: -1000vh; box-sizing: border-box;"
  >
    <div>
      <div
        class="rc-cascader-menus"
      >
        <ul
          class="rc-cascader-menu"
          role="menu"
        >
          <li
            aria-checked="false"
            class="rc-cascader-menu-item"
            data-path-key="Zhejiang__RC_CASCADER_SPLIT__Hangzhou__RC_CASCADER_SPLIT__West Lake"
            role="menuitemcheckbox"
            title="Zhejiang / Hangzhou / West Lake"
          >
            <div
              class="rc-cascader-menu-item-content"
            >
              Zhejiang / Hangzhou / West Lake
            </div>
          </li>
          <li
            aria-checked="false"
            class="rc-cascader-menu-item rc-cascader-menu-item-disabled"
            data-path-key="Zhejiang__RC_CASCADER_SPLIT__Hangzhou__RC_CASCADER_SPLIT__Xia Sha"
            role="menuitemcheckbox"
            title="Zhejiang / Hangzhou / Xia Sha"
          >
            <div
              class="rc-cascader-menu-item-content"
            >
              Zhejiang / Hangzhou / Xia Sha
            </div>
          </li>
        </ul>
      </div>
    </div>
  </div>
</div>
`;
